src/silvimetric/resources/metric.py

```diff
@@ -48,39 +55,104 @@ def entry_name(self, attr: str) -> str:
         """Name for use in TileDB and extract file generation."""
         return f'm_{attr}_{self.name}'
 
+    def do(self, data: pd.DataFrame) -> pd.DataFrame:
+        """Run metric and filters."""
+        if self.attributes:
+            attrs = [*[a.name for a in self.attributes],'xi','yi']
+            data = data[attrs]
+
+        data = self.run_filters(data)
+        idx = ['xi','yi']
+        gb = data.groupby(idx)
+        cols = data.columns
+        new_cols = {
+            c: [(self.entry_name(c), self._method)]
+            for c in cols if c not in idx
+        }
+
+        val= gb.agg(new_cols)
+        #remove hierarchical columns
+        val.columns = val.columns.droplevel(0)
+        return val
+
     @dask.delayed
-    def delayed(self, data: np.ndarray) -> np.ndarray:
-        return self._method(data)
+    def do_delayed(self, data: pd.DataFrame) -> pd.DataFrame:
+        """Run metric as a dask delayed method"""
+        return self.do(data)
+
+    #TODO make dict with key for each Attribute effected? {att: [fn]}
+    # for now these filters apply to all Attributes
+    def add_filter(self, fn: FilterFn, desc: str):
+        """
+        Add filter method to list of filters to run before calling main method.
+        """
+        self.filters.append(fn)
+
+    def run_filters(self, data: pd.DataFrame) -> pd.DataFrame:
+        for f in self.filters:
+            ndf = f(data)
+            if not isinstance(ndf, pd.DataFrame):
```

Comment:
No need to fix this now but isn't it good enough to *act* like a DataFrame here?

Reply:
It probably is, and there is definitely a line here where we just let the user figure it out, but it would be a lot less error prone (though also maybe less flexible) if we have a check in between each stage that the product is coming out correctly.